### PR TITLE
feature(pu): add efficientzero and related modules

### PR DESCRIPTION
**Pull request by puyuan1996**
add efficientzero policy and model, and related env and config, and migrate the existing alphazero demo

## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/model/template/efficientzero/efficientzero_base_model.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+import torch
+import typing
+
+import numpy as np
+import torch.nn as nn
+
+from typing import List
+
+
+class NetworkOutput(typing.NamedTuple):
```
Comment (PaParaZz1) [inline ding/model/template/efficientzero/efficientzero_base_model.py:14]: move to dataclass, you can refer to [this implementation](https://github.com/opendilab/DI-engine/blob/main/ding/data/buffer/buffer.py#L41)

File: ding/model/template/efficientzero/efficientzero_base_model.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    def dynamics(self, state, reward_hidden, action):
+        raise NotImplementedError
+
+    def initial_inference(self, obs) -> NetworkOutput:
+        num = obs.size(0)
+
+        state = self.representation(obs)
+        actor_logit, value = self.prediction(state)
+
+        if not self.training:
```
Comment (PaParaZz1) [inline ding/model/template/efficientzero/efficientzero_base_model.py:93]:
```python
if self.training:
  ...
else:
  ...
```

File: ding/model/template/efficientzero/efficientzero_base_model.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            # if not in training, obtain the scalars of the value/reward
+            # tictactoe: value {Tensor: (2,32)}
+            value = self.inverse_value_transform(value).detach().cpu().numpy()
```
Comment (PaParaZz1) [inline ding/model/template/efficientzero/efficientzero_base_model.py:96]: if not self.training, should we remove line 90-91 in this branch?

File: ding/model/template/efficientzero/efficientzero_base_model.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            # if not in training, obtain the scalars of the value/reward
+            value = self.inverse_value_transform(value).detach().cpu().numpy()
+            value_prefix = self.inverse_reward_transform(value_prefix).detach().cpu().numpy()
+            state = state.detach().cpu().numpy()
+            reward_hidden = (reward_hidden[0].detach().cpu().numpy(), reward_hidden[1].detach().cpu().numpy())
+            actor_logit = actor_logit.detach().cpu().numpy()
+
+        return NetworkOutput(value, value_prefix, actor_logit, state, reward_hidden)
+
+    def get_weights(self):
```
Comment (PaParaZz1) [inline ding/model/template/efficientzero/efficientzero_base_model.py:125]: unify to `state_dict` and `load_state_dict` methods

File: ding/model/template/efficientzero/efficientzero_base_model.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        return {k: v.cpu() for k, v in self.state_dict().items()}
```
Comment (PaParaZz1) [inline ding/model/template/efficientzero/efficientzero_base_model.py:126]: use ordereddict

File: ding/model/template/efficientzero/efficientzero_base_model.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+    def set_weights(self, weights):
+        self.load_state_dict(weights)
+
+    def get_gradients(self):
+        grads = []
+        for p in self.parameters():
+            grad = None if p.grad is None else p.grad.data.cpu().numpy()
```
Comment (PaParaZz1) [inline ding/model/template/efficientzero/efficientzero_base_model.py:134]: don't use any numpy data in model

File: ding/model/template/efficientzero/efficientzero_tictactoe_model.py [hunk 7]
```diff
@@ -0,0 +1,607 @@
...
+    return inverse_scalar_transform(reward_logits, reward_support)
+    # return reward_logits  # TODO
+
+
+def inverse_value_transform(value_logits):
+    return inverse_scalar_transform(value_logits, value_support)
+    # return value_logits  # TODO
+
+
+def inverse_scalar_transform(logits, scalar_support):
```
Comment (PaParaZz1) [inline ding/model/template/efficientzero/efficientzero_tictactoe_model.py:39]: can we divide this function into two parts: `C51Head` in model and `value_rescale` function in policy

File: ding/model/template/efficientzero/efficientzero_tictactoe_model.py [hunk 8]
```diff
@@ -0,0 +1,607 @@
...
+    value_support = value_support.to(device=value_probs.device)
+    value = (value_support * value_probs).sum(1, keepdim=True) / delta
+
+    epsilon = 0.001
+    sign = torch.ones(value.shape).float().to(value.device)
+    sign[value < 0] = -1.0
+    output = (((torch.sqrt(1 + 4 * epsilon * (torch.abs(value) + 1 + epsilon)) - 1) / (2 * epsilon)) ** 2 - 1)
+    output = sign * output * delta
+
+    nan_part = torch.isnan(output)
```
Comment (PaParaZz1) [inline ding/model/template/efficientzero/efficientzero_tictactoe_model.py:58]: abstract a general function to mask nan part

File: ding/model/template/efficientzero/efficientzero_tictactoe_model.py [hunk 9]
```diff
@@ -0,0 +1,607 @@
...
+    output[nan_part] = 0.
+    output[torch.abs(output) < epsilon] = 0.
+    return output
+
+
+# value=torch.randn([2,32])
+# inverse_value_transform(value)
+
+
+def mlp(
```
Comment (PaParaZz1) [inline ding/model/template/efficientzero/efficientzero_tictactoe_model.py:68]: why not use our MLP here

File: ding/model/template/efficientzero/efficientzero_tictactoe_model.py [hunk 10]
```diff
@@ -0,0 +1,607 @@
...
+# fc = mlp(27, [64,32,16,16], 9)
+# print(fc)
+
+
+def conv3x3(in_channels, out_channels, stride=1):
+    return nn.Conv2d(in_channels, out_channels, kernel_size=3, stride=stride, padding=1, bias=False)
+
+
+# Residual block
+class ResidualBlock(nn.Module):
```
Comment (PaParaZz1) [inline ding/model/template/efficientzero/efficientzero_tictactoe_model.py:118]: use our residual block implementation, if there are some mismatch parts, modify our codes adapted to the demand of muzero/efficientzero

File: ding/model/template/efficientzero/efficientzero_tictactoe_model.py [hunk 11]
```diff
@@ -0,0 +1,607 @@
...
+class RepresentationNetwork(nn.Module):
+
+    def __init__(self, ):
+        """
+        Representation network
+        equivalence transformation
+        """
+        super().__init__()
+        # be compatiable with method get_param_mean
+        self._dummy_param = nn.Parameter(torch.zeros(1, 1))
```
Comment (PaParaZz1) [inline ding/model/template/efficientzero/efficientzero_tictactoe_model.py:209]: why dummy_param here

File: ding/model/template/efficientzero/efficientzero_tictactoe_model.py [hunk 12]
```diff
@@ -0,0 +1,607 @@
...
+        x += state
+        x = nn.functional.relu(x)
+
+        for block in self.resblocks:
+            x = block(x)
+        state = x
+
+        x = self.conv1x1_reward(x)
+        x = self.bn_reward(x)
+        x = nn.functional.relu(x)
```
Comment (PaParaZz1) [inline ding/model/template/efficientzero/efficientzero_tictactoe_model.py:292]: use torch.nn as last activation

File: ding/model/template/efficientzero/efficientzero_tictactoe_model.py [hunk 13]
```diff
@@ -0,0 +1,607 @@
...
+    def project(self, hidden_state, with_grad=True):
+        # only the branch of proj + pred can share the gradients
+        # hidden_state = hidden_state.view(-1, self.porjection_in_dim)
+        hidden_state = hidden_state.reshape(-1, self.porjection_in_dim)
+
+        proj = self.projection(hidden_state)
+
+        # with grad, use proj_head
+        if with_grad:
+            proj = self.projection_head(proj)
```
Comment (PaParaZz1) [inline ding/model/template/efficientzero/efficientzero_tictactoe_model.py:604]: `return self.projection_head(proj)`

File: ding/model/template/efficientzero/efficientzero_tictactoe_model.py [hunk 14]
```diff
@@ -0,0 +1,607 @@
...
+        encoded_state = self.representation_network(observation)
+        if not self.state_norm:
+            return encoded_state
+        else:
+            encoded_state_normalized = renormalize(encoded_state)
+            return encoded_state_normalized
+
+    def dynamics(self, encoded_state, reward_hidden, action):
+        # Stack encoded_state with a game specific one hot encoded action
+        action_one_hot = (
```
Comment (PaParaZz1) [inline ding/model/template/efficientzero/efficientzero_tictactoe_model.py:570]: use our `one_hot` function

File: ding/model/template/efficientzero/efficientzero_tictactoe_model.py [hunk 15]
```diff
@@ -0,0 +1,607 @@
+"""
```
Comment (PaParaZz1) [inline ding/model/template/efficientzero/efficientzero_tictactoe_model.py:1]: remove gomoku related codes like model and config now

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/model/template/efficientzero/efficientzero_base_model.py (hunk 3 above)
Comment (puyuan1996) [reply ding/model/template/efficientzero/efficientzero_base_model.py:96]: 无论是否 self.training，应该都需要计算 state,value,policy_logits

File: ding/model/template/efficientzero/efficientzero_tictactoe_model.py (hunk 14 above)
Comment (puyuan1996) [reply ding/model/template/efficientzero/efficientzero_tictactoe_model.py:570]: 这里的one-hot不是通常的形式，是shape为(1,3,3), 所有元素都为action*1/action_space_size 这样的

File: ding/model/template/efficientzero/efficientzero_tictactoe_model.py (hunk 15 above)
Comment (puyuan1996) [reply ding/model/template/efficientzero/efficientzero_tictactoe_model.py:1]: ok

File: ding/rl_utils/efficientzero/check/check_mcts.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
+import os
```
Comment (PaParaZz1) [inline ding/rl_utils/efficientzero/check/check_mcts.py:1]: remove `check` now

File: ding/rl_utils/efficientzero/dataset.py [hunk 17]
```diff
@@ -0,0 +1,64 @@
+"""
+The following code is adapted from https://github.com/YeWR/EfficientZero/core/dataset.py
+"""
+
+import torch
+import torch.nn as nn
+
+from kornia.augmentation import RandomAffine, RandomCrop, CenterCrop, RandomResizedCrop
```
Comment (PaParaZz1) [inline ding/rl_utils/efficientzero/dataset.py:8]: can we use torch official transform and not include new lib requirement?

File: ding/rl_utils/efficientzero/dataset.py [hunk 18]
```diff
@@ -0,0 +1,64 @@
+"""
```
Comment (PaParaZz1) [inline ding/rl_utils/efficientzero/dataset.py:1]: move to `image_transform_muzero.py`

File: ding/rl_utils/muzero/check_mcts.py [hunk 19]
```diff
@@ -0,0 +1,102 @@
+import torch
```
Comment (PaParaZz1) [inline ding/rl_utils/muzero/check_mcts.py:1]: remove thisfile and related code

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/model/template/efficientzero/efficientzero_base_model.py [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+
+    return output
+
+
+def renormalize(tensor, first_dim=1):
+    # normalize the tensor (states)
+    if first_dim < 0:
+        first_dim = len(tensor.shape) + first_dim
+    flat_tensor = tensor.view(*tensor.shape[:first_dim], -1)
+    max = torch.max(flat_tensor, first_dim, keepdim=True).values
```
Comment (sheng-han-zhang) [inline ding/model/template/efficientzero/efficientzero_base_model.py:111]: max/min 这个命名换一下吧，和默认函数重合，max_val/min_val

File: ding/model/template/efficientzero/efficientzero_base_model.py [hunk 21]
```diff
@@ -0,0 +1,115 @@
+"""
+The following code is adapted from https://github.com/YeWR/EfficientZero/blob/main/core/model.py
+"""
+
+import torch
+import numpy as np
+import torch.nn as nn
+from typing import List, NamedTuple
+from dataclasses import dataclass
+from ding.rl_utils.mcts.utils import mask_nan
```
Comment (sheng-han-zhang) [inline ding/model/template/efficientzero/efficientzero_base_model.py:10]: 这个mask_nan有必要么？假设tensor里有nan，感觉不大好。

File: ding/model/template/efficientzero/efficientzero_base_model.py [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+    scalar_support = DiscreteSupport(-support_size, support_size, delta=1)
+    delta = scalar_support.delta
+    value_probs = torch.softmax(logits, dim=1)
+    value_support = torch.ones(value_probs.shape)
+    value_support[:, :] = torch.from_numpy(np.array([x for x in scalar_support.range]))
+    value_support = value_support.to(device=value_probs.device)
+    value = (value_support * value_probs).sum(1, keepdim=True) / delta
+
+    epsilon = 0.001
+    sign = torch.ones(value.shape).float().to(value.device)
```
[the remark this answers is not in the document]
Comment (puyuan1996) [reply ding/model/template/efficientzero/efficientzero_base_model.py:97]: 好的，谢谢啦

File: ding/model/template/efficientzero/efficientzero_base_model.py (hunk 21 above)
Comment (puyuan1996) [reply ding/model/template/efficientzero/efficientzero_base_model.py:10]: 这个是之前ez仓库这么写的，之后我改一下

File: ding/model/template/efficientzero/efficientzero_base_model.py (hunk 20 above)
Comment (puyuan1996) [reply ding/model/template/efficientzero/efficientzero_base_model.py:112]: 好的

File: ding/model/template/tests/test_efficientzero_tictactoe_model.py [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+        repnet = RepresentationNetwork()
+        state = repnet(obs)
+        assert state.size() == obs.size()
+
+    @pytest.mark.parametrize(
+        'num_blocks, num_channels, reduced_channels_reward, fc_reward_layers, full_support_size,'
+        'block_output_size_reward',
+        dyn_args
+    )
+    def test_DynamicsNetwork(
```
Comment (PaParaZz1) [inline ding/model/template/tests/test_efficientzero_tictactoe_model.py:49]: polish name

File: ding/model/template/tests/test_efficientzero_tictactoe_model.py [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+        self, num_blocks, num_channels, reduced_channels_reward, fc_reward_layers, full_support_size,
+        block_output_size_reward
+    ):
+        batch = 100  # this is (torch.randn(1, 10, 64), torch.randn(1, 10, 64)) => 100 / 10 = 10
+        state = torch.rand(batch, 3, 3, 3)
+        dynnet = DynamicsNetwork(
```
Comment (PaParaZz1) [inline ding/model/template/tests/test_efficientzero_tictactoe_model.py:55]: typo

File: ding/model/template/tests/test_efficientzero_tictactoe_model.py [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+            num_blocks=num_blocks,
+            num_channels=num_channels,
+            reduced_channels_reward=reduced_channels_reward,
+            fc_reward_layers=fc_reward_layers,
+            full_support_size=full_support_size,
+            block_output_size_reward=block_output_size_reward
+        )
+        state_, reward_hidden, value_prefix = dynnet(state, (torch.randn(1, 10, 64), torch.randn(1, 10, 64)))
```
Comment (PaParaZz1) [inline ding/model/template/tests/test_efficientzero_tictactoe_model.py:63]: why only assert state_

File: ding/model/template/efficientzero/efficientzero_model.py [hunk 26]
```diff
@@ -0,0 +1,573 @@
...
+            x = block(x)
+        x = self.pooling1(x)
+        for block in self.resblocks3:
+            x = block(x)
+        x = self.pooling2(x)
+        return x
+
+
+# Encode the observations into hidden states
+class RepresentationNetworkIndIdentity(nn.Module):
```
Comment (PaParaZz1) [inline ding/model/template/efficientzero/efficientzero_model.py:107]: use `nn.Identity()`

File: ding/worker/collector/base_serial_evaluator_muzero.py [hunk 27]
```diff
@@ -0,0 +1,448 @@
...
+import torch
+
+from ding.utils import build_logger, EasyTimer, lists_to_dicts
+from ding.envs import BaseEnvManager
+from ding.torch_utils import to_tensor, to_ndarray, tensor_to_list
+from ding.rl_utils.mcts.game import GameHistory
+from ding.rl_utils.mcts.utils import prepare_observation_lst
+
+
+class BaseSerialEvaluatorMuZero(object):
```
Comment (PaParaZz1) [inline ding/worker/collector/base_serial_evaluator_muzero.py:15]: MuZeroEvaluator

Comment (PaParaZz1) [inline ding/worker/collector/base_serial_evaluator_muzero.py:15]: You can extend `BaseSerialEvaluator` to reuse some methods

File: ding/worker/collector/base_serial_evaluator_muzero.py [hunk 28]
```diff
@@ -0,0 +1,448 @@
...
+        if stop_flag:
+            self._logger.info(
+                "[DI-engine serial pipeline] " +
+                "Current eval_reward: {} is greater than stop_value: {}".format(eval_reward, self._stop_value) +
+                ", so your RL agent is converged, you can refer to 'log/evaluator/evaluator_logger.txt' for details."
+            )
+        return stop_flag, eval_reward
+
+
+class VectorEvalMonitor(object):
```
Comment (PaParaZz1) [inline ding/worker/collector/base_serial_evaluator_muzero.py:347]: don't use repeat codes

File: ding/worker/collector/base_serial_evaluator_muzero.py [hunk 29]
```diff
@@ -0,0 +1,448 @@
...
+                        eval_monitor.update_reward(env_id, reward)
+                        self._logger.info(
+                            "[EVALUATOR]env {} finish episode, final reward: {}, current episode: {}".format(
+                                env_id, eval_monitor.get_latest_reward(env_id), eval_monitor.get_current_episode()
+                            )
+                        )
+
+                        # reset the finished env
+                        init_obses = self._env.ready_obs
+                        if two_plaer_game:
```
Comment (PaParaZz1) [inline ding/worker/collector/base_serial_evaluator_muzero.py:288]: typo

File: ding/worker/collector/base_serial_evaluator_muzero.py [hunk 30]
```diff
@@ -0,0 +1,448 @@
...
+        action_mask = [init_obses[i]['action_mask'] for i in range(env_nums)]
+        if 'to_play' in init_obses[0]:
+            two_plaer_game = True
+        else:
+            two_plaer_game = False
+        if two_plaer_game:
+            to_play = [init_obses[i]['to_play'] for i in range(env_nums)]
+        dones = np.array([False for _ in range(env_nums)])
+
+        game_histories = [
```
Comment (PaParaZz1) [inline ding/worker/collector/base_serial_evaluator_muzero.py:214]: can we remote `game_histories` in evaluator?

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/torch_utils/network/nn_module.py [hunk 31]
```diff
@@ -302,7 +302,8 @@ def MLP(
     activation: nn.Module = None,
     norm_type: str = None,
     use_dropout: bool = False,
-    dropout_probability: float = 0.5
+    dropout_probability: float = 0.5,
+    init_zero: bool = False,
```
[the remark this answers is not in the document]
Comment (puyuan1996) [reply ding/torch_utils/network/nn_module.py:306]: 为了与EZ原仓库一致，将这个扩展为可以单独指定最后一层的norm activation init

File: ding/model/template/tests/test_efficientzero_tictactoe_model.py (hunk 23 above)
Comment (puyuan1996) [reply ding/model/template/tests/test_efficientzero_tictactoe_model.py:49]: ok

File: ding/model/template/tests/test_efficientzero_tictactoe_model.py (hunk 25 above)
Comment (puyuan1996) [reply ding/model/template/tests/test_efficientzero_tictactoe_model.py:63]: added others

File: ding/worker/collector/base_serial_evaluator_muzero.py (hunk 28 above)
Comment (puyuan1996) [reply ding/worker/collector/base_serial_evaluator_muzero.py:347]: ok ,thanks

**Pull request closed without merging**